Tolerations and affinity

totycro opened the pull request:
This PR adds tolerations and affinity settings for all pods. This is necessary for clusters with multiple node types, where the operators want to run eoapi on specific ones.

I've tested the change with the following values. With this all pods run on the desired nodes (except for pgbounce and pgbackuprest, but these can be [configured by using `postgrescluster` values](https://access.crunchydata.com/documentation/postgres-operator/latest/references/crd/5.7.x/postgrescluster#postgresclusterspecproxypgbounceraffinity)).

The values are a bit repetitive, but it might be a realistic use case that e.g. titiler should run on specific nodes, which would not be possible with a global tolerations/affinity setting (although it would be enough for my use case for now).

I'm not sure if these settings can realistically be tested in CI in k3s. We'd need to have different nodes in k3s and make sure that the services are spawned in the correct ones.

Related discussion: https://github.com/developmentseed/eoapi-k8s/discussions/172

```
stac:
  settings:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user
    tolerations:
      - key: "my-taint-name"
        operator: "Equal"
        value: "user"
        effect: "NoSchedule"

raster:
  settings:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user
    tolerations:
      - key: "my-taint-name"
        operator: "Equal"
        value: "user"
        effect: "NoSchedule"

vector:
  settings:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user
    tolerations:
      - key: "my-taint-name"
        operator: "Equal"
        value: "user"
        effect: "NoSchedule"

docServer:
  settings:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user
    tolerations:
      - key: "my-taint-name"
        operator: "Equal"
        value: "user"
        effect: "NoSchedule"

pgstacBootstrap:
  settings:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user
    tolerations:
      - key: "my-taint-name"
        operator: "Equal"
        value: "user"
        effect: "NoSchedule"

postgrescluster:
  instances:
  - name: eoapi
    replicas: 1
    affinity: 
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: "my-node-key"
              operator: In
              values:
              - user

    tolerations: 
    - key: "my-taint-name"
      operator: "Equal"
      value: "user"
      effect: "NoSchedule"

    dataVolumeClaimSpec:
      accessModes:
      - "ReadWriteOnce"
      resources:
        requests:
          storage: "10Gi"
          cpu: "1024m"
          memory: "3048Mi"
```

pantierra commented on helm-chart/eoapi/templates/services/deployment.yaml:
```diff
@@ -71,6 +71,14 @@ spec:
             name: {{ $secret }}
         {{- end }}
         {{- end }}
+      {{- with index $v "settings" "affinity" }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with index $v "settings" "tolerations" }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
Can we check for the existence of `affinity` and `tolerations`, and possibly avoid adding them as empty objects to `values.yaml`?

```suggestion
        {{- if index $v "settings" "affinity" }}
        affinity:
          {{- toYaml (index $v "settings" "affinity") | nindent 8 }}
        {{- end }}
        {{- if index $v "settings" "tolerations" }}
        tolerations:
          {{- toYaml (index $v "settings" "tolerations") | nindent 8 }}
        {{- end }}
```

pantierra commented on helm-chart/eoapi/templates/pgstacboostrap/job.yaml:
```diff
@@ -51,5 +51,13 @@ spec:
         - name: pgstac-setup-volume-{{ $.Release.Name }}
           configMap:
             name: pgstac-setup-config-{{ $.Release.Name }}
+      {{- with .Values.pgstacBootstrap.settings.affinity }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .Values.pgstacBootstrap.settings.tolerations }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
Can we check for the existence of `affinity` and `tolerations`, and possibly avoid adding them as empty objects to `values.yaml`?

```suggestion
      {{- if .Values.pgstacBootstrap.settings.affinity }}
      affinity:
        {{- toYaml .Values.pgstacBootstrap.settings.affinity | nindent 8 }}
      {{- end }}
      {{- if .Values.pgstacBootstrap.settings.tolerations }}
      tolerations:
        {{- toYaml .Values.pgstacBootstrap.settings.tolerations | nindent 8 }}
      {{- end }}
```

pantierra commented on helm-chart/eoapi/templates/services/nginx-doc-server.yaml:
```diff
@@ -46,6 +46,14 @@ spec:
       - name: doc-html-{{ .Release.Name }}
         configMap:
           name: nginx-root-html-{{ .Release.Name }}
+      {{- with .Values.docServer.settings.affinity }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .Values.docServer.settings.tolerations }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
Can we check for the existence of `affinity` and `tolerations`, and possibly avoid adding them as empty objects to `values.yaml`?

```suggestion
        {{- if .Values.docServer.settings }}
        {{- if .Values.docServer.settings.affinity }}
        affinity:
          {{- toYaml .Values.docServer.settings.affinity | nindent 8 }}
        {{- end }}
        {{- if .Values.docServer.settings.tolerations }}
        tolerations:
          {{- toYaml .Values.docServer.settings.tolerations | nindent 8 }}
        {{- end }}
        {{- end }}
```

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -108,6 +108,8 @@ postgrescluster:
   instances:
   - name: eoapi
     replicas: 1
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -165,6 +167,8 @@ pgstacBootstrap:
       LOAD_FIXTURES: "true"
       # toggle to `true` if you want to keep the job running as db jumpbox
       KEEP_ALIVE: "false"
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -230,6 +234,8 @@ raster:
       PORT: "8080"
       # https://www.uvicorn.org/settings/#production
       WEB_CONCURRENCY: "5"
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -279,6 +285,8 @@ stac:
       PORT: "8080"
       # https://www.uvicorn.org/settings/#production
       WEB_CONCURRENCY: "5"
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -333,6 +341,11 @@ vector:
       PORT: "8080"
       # https://www.uvicorn.org/settings/#production
       WEB_CONCURRENCY: "5"
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra commented on helm-chart/eoapi/values.yaml:
```diff
@@ -333,6 +341,11 @@ vector:
 
 docServer:
   enabled: true
+  settings:
+    affinity: {}
+    tolerations: []
```
Can they be removed?

```suggestion
````

pantierra reviewed (commented):
Many thanks for the PR! Looks good, already. I was just wondering whether we could avoid setting empty values for `affinity` and `tolerations` in the `values.yaml`. I'd prefer to have them added only if needed.

totycro commented:
> Many thanks for the PR! Looks good, already. I was just wondering whether we could avoid setting empty values for `affinity` and `tolerations` in the `values.yaml`. I'd prefer to have them added only if needed.

Thanks for the review @pantierra!

I've removed the empty values from values.yaml as you have suggested. I've added them because then they have more visibility, but it would also work if they are mentioned in somewhere else.

Btw I used the `with` syntax because it has in implicit `if`, so no output would be generated if the affinity/tolerations are not present or an empty list or an empty object. Also then you only need to specify the key once since it updates the dot to refer to the value.
Is there a particular reason why you prefer an explicit `if` here? It would mean a bit less helm template magic and more consistency with the rest.

pantierra commented:
No, all good. `with` is just fine then!
